homeassistant/components/filter/sensor.py

```diff
@@ -330,7 +330,7 @@ def __init__(self, state):
         self.timestamp = state.last_updated
         try:
             self.state = float(state.state)
-        except ValueError:
+        except (TypeError, ValueError):
```

Comment:
`state.state` cannot be `None` ?

Reply:
in the linked issue it is None :(

Reply:
Then you have a bug somewhere else. 

We always cast state to a string.

https://github.com/home-assistant/core/blob/819dd27925c0cbcb45c85ce81c32746538d6a6fa/homeassistant/core.py#L854

Reply:
guess it could come from https://github.com/home-assistant/core/blob/68221907729405836890e56a879e363b1c003510/homeassistant/components/history/__init__.py#L201

Should I cast it there to a string ?

Reply:
Well that's not a HA state object 🤔 . Adding types to this code should help find it.

Reply:
tracking the error I found out that the None can be generated when an entity is reloaded:

```
2020-11-21 01:12:45 ERROR (MainThread) [homeassistant.components.filter.sensor] <Event state_changed[L]: entity_id=sensor.teste, old_state=<state sensor.teste=1000; friendly_name=teste @ 2020-11-21T01:12:39.964030+00:00>, new_state=None>
2020-11-21 01:12:45 ERROR (MainThread) [homeassistant.helpers.event] Error while processing state changed for sensor.teste
Traceback (most recent call last):
  File "/Users/dgomes/Projects/home-assistant/homeassistant/helpers/event.py", line 257, in _async_state_change_dispatcher
    hass.async_run_hass_job(job, event)
  File "/Users/dgomes/Projects/home-assistant/homeassistant/core.py", line 431, in async_run_hass_job
    hassjob.target(*args)
  File "/Users/dgomes/Projects/home-assistant/homeassistant/components/filter/sensor.py", line 186, in _update_filter_sensor_state_event
    self._update_filter_sensor_state(event.data.get("new_state"))
  File "/Users/dgomes/Projects/home-assistant/homeassistant/components/filter/sensor.py", line 192, in _update_filter_sensor_state
    self._state = new_state.state
AttributeError: 'NoneType' object has no attribute 'state'
2020-11-21 01:12:45 ERROR (MainThread) [homeassistant.components.filter.sensor] <Event state_changed[L]: entity_id=sensor.teste, old_state=None, new_state=<state sensor.teste=1000; friendly_name=teste @ 2020-11-21T01:12:45.109897+00:00>>
```

In this case I have a filter over a very simple template sensor `{{ 1000 }}`

Reply:
I've opened a new issue, this has nothing to do with history, its event change

Reply:
Actually got a test that shows that when template is reloaded the all "State" becomes None:

https://github.com/home-assistant/core/pull/44439/files#diff-979128f340f102ef3e390c796f36715b5e5c713a4965f4d4ccba42f932fc9befR139

Reply:
In a more recent Issue: https://github.com/home-assistant/core/issues/44424

The log shows the `None` comes from the history component `(None, datetime.datetime(2020, 12, 22, 12, 33, 27, 677473, tzinfo=<UTC>))`

Reply:
When you query the `history` integration it does not return `State` but a `LazyState` https://github.com/home-assistant/core/blob/68221907729405836890e56a879e363b1c003510/homeassistant/components/history/__init__.py#L338

And LazyState can in fact be None...

So can we move this forward :) ?